contributor-book/src/project-architecture/serialization.md

```diff
@@ -0,0 +1,101 @@
+# Serialization
+
+An important aspect of a deep learning framework is the ability to save and load models from disk.
+Despite appearing as a simple feature, it involves numerous constraints that require a proper
+solution.
+
+## Constraints
+
+1. **Users should be able to declare the precision of the model to be saved, independent of the
+   backend in use.**
+
+   The modules should not be duplicated in RAM in another precision to support this. Conversion
+   should be done lazily during (de)serialization.
+
+2. **Users should be able to add any field to a module, even fields that are not serializable.**
+
+   This can include constants, database connections, other module references, or any other
+   information. Only parameters should be serialized since the structure of the module itself should
+   be encapsulated with module configurations (hyper-parameters).
+
+3. **Users should be able to declare the format in which the module should be saved.**
+
+   This can involve saving to a compressed JSON file or directly to bytes in memory for `no-std`
+   environments.
+
+4. **Users should be able to create a module with its saved parameters without having to initialize
+   the module first.**
+
+   This will avoid unnecessary module initialization and tensor loading, resulting in reduced cold
+   start when dealing with inference.
+
+In addition to all of these constraints, the solution should be easy to use.
+
+## Solution
+
+In order to be able to add any field to a module without requiring it to be (de)serializable, we
+decouple the module type from its state. We create a new type for each module that only contains the
+parameters that need to be saved. To generate that type automatically, the user must either declare
+which field is a parameter or a constant, or we assume that each field implements the module trait.
+
+The second solution was chosen as it simplifies the code generation and reduces the size of the user
+API. This means that the `Module` trait should be implemented by
+[primitives types](./burn-core/src/module/param/primitive.rs). The following diagrams highlight the
+main types and traits used in the solution.
+
+<div align="center">
+<h4>Module Serialization Types</h4>
+<img src="./assets/ModuleSerialization.png" width="700px"/>
```

Comment:
The image doesn't render.

Reply:
copied the image to the directory, we may want to delete the old one if we also delete the old architecture.md (which is the same as architecture split into different sections)